src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/linalg/FourierLinearOperatorNDArray.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+
+public final class FourierLinearOperatorNDArray extends FourierLinearOperator<INDArray> {
+
+    private final double[] orderedFourierFactors;
+
+    public FourierLinearOperatorNDArray(final int dimension, final double[] fourierFactors,
+                                        final boolean promoteToPowerOfTwo) {
+        super(dimension, fourierFactors, promoteToPowerOfTwo);
+        orderedFourierFactors = new double[fftSize];
+        initializeOrderedFourierFactors();
+    }
+
+    /**
+     * Implementation of the action of the Fourier linear operator on a given real vector
+     * @param x an {@link INDArray} on which the linear operator is acted on
+     * @return the transformed vector
+     */
+    @Override
+    public INDArray operate(@Nonnull final INDArray x) throws DimensionMismatchException {
+        if (x.length() != dimension) {
+            throw new DimensionMismatchException(x.length(), dimension);
+        }
+        final double[] xDoubleArray = createDoubleArrayFromINDArray(x);
+        performForwardFFTInPlace(xDoubleArray);
+        applyFilterInPlace(xDoubleArray);
+        performInverseFFTInPlace(xDoubleArray);
+        return Nd4j.create(xDoubleArray, new int[]{dimension, 1});
+    }
+
+    /**
+     * @param x the input vector
+     * @return rfft of x
+     */
+    public double[] getForwardFFT(@Nonnull final INDArray x) {
+        if (x.length() != dimension) {
+            throw new DimensionMismatchException(x.length(), dimension);
+        }
+        final double[] xDoubleArray = createDoubleArrayFromINDArray(x);
+        performForwardFFTInPlace(xDoubleArray);
+        return xDoubleArray;
+    }
+
+    /**
+     * @param xFFT the input vector
+     * @return rfft of x
+     */
+    public INDArray getInverseFFT(@Nonnull final double[] xFFT) {
+        if (xFFT.length != fftSize) {
+            throw new DimensionMismatchException(xFFT.length, fftSize);
+        }
+        final double[] xDoubleArray = xFFT.clone();
+        performInverseFFTInPlace(xDoubleArray);
+        return Nd4j.create(xDoubleArray, new int[]{dimension, 1});
+    }
+
+    /**
+     * @param xFFT the input vector
+     * @return rfft of x
+     */
+    public INDArray getInverseFFT(@Nonnull final INDArray xFFT) {
+        if (xFFT.length() != fftSize) {
+            throw new DimensionMismatchException(xFFT.length(), fftSize);
+        }
+        final double[] xFFTDoubleArray = new double[fftSize];
+        for (int i = 0; i < fftSize; i++) {
+            xFFTDoubleArray[i] = xFFT.getDouble(i);
+        }
+        performInverseFFTInPlace(xFFTDoubleArray);
+        return Nd4j.create(xFFTDoubleArray, new int[]{dimension, 1});
+    }
+
+    public double[] getOrderedFourierFactors() {
+        return orderedFourierFactors.clone();
+    }
+
+    private void initializeOrderedFourierFactors() {
```

Comment:
This could use some comments/documentation to make clear the definition of the orderedFourierFactors.